src/frame/endpoints/UserApiEndpoints.ts

```diff
@@ -307,7 +323,28 @@ export default {
         if (!userList.length) {
             return Future.of({result: []});
         }
-        const {url, options, errorCode} = userKeyList(userList);
-        return makeAuthorizedApiRequest(url, errorCode, options);
+
+        // Check the list for any users we already have the public key cached for
+        const cacheHits: UserKeyListResponseObject[] = [];
+        userList.forEach((userId) => {
+            const userMasterPublicKey = userPublicKeyCache[userId];
+            if (userMasterPublicKey) {
+                cacheHits.push({id: userId, ...userMasterPublicKey});
+            }
+        });
+
+        // If we have any userIds that weren't cached, fetch all of them
+        if (cacheHits.length === userList.length) {
+            return Future.of({result: cacheHits});
+        } else {
+            const {url, options, errorCode} = userKeyList(userList);
+            return makeAuthorizedApiRequest<UserKeyListResponseType>(url, errorCode, options).map((userListResponse) => {
+                // cache the retrieved keys
+                userListResponse.result.forEach((userKey) => {
+                    userPublicKeyCache[userKey.id] = userKey;
```

Comment:
Isn't `userKey` an object with an ID and a public key? This seems like it shouldn't compile, since the cache is supposed to map from a string to a `PublicKey`.

Reply:
Ah, it's cause of TS allowing flexible structural typing. The cache is `{ [id: string]: { userMasterPublicKey: string } }` and what's being put into it is `{ [id: string]: { id: string, userMasterPublicKey: string } }`. Typescript just sees that the object being placed _at least_ contains everything that the type requires, so no code would break, and says it's fine. I think there's a way to tell it to be more strict on specific types, but good catch, I'll fix this.